File: examples/governance/model_registry_domain/lib/model_registry_domain/data_usage_agreement/commands/create_agreement.rb
module ModelRegistryDomain
  class DataUsageAgreement
    module Commands
      class CreateAgreement
        include Hecks::Command
        emits "CreatedAgreement"

        attr_reader :model_id
        attr_reader :data_source
        attr_reader :purpose
        attr_reader :consent_type

        def initialize(
          model_id: nil,
          data_source: nil,
          purpose: nil,
          consent_type: nil
        )
          @model_id = model_id
          @data_source = data_source
          @purpose = purpose
          @consent_type = consent_type
        end

        def call
          DataUsageAgreement.new(
            model_id: model_id,
            data_source: data_source,
            purpose: purpose,
            consent_type: consent_type,
            status: "draft"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/create_agreement.go
package domain

import (
	"time"
)

type CreateAgreement struct {
	ModelId string `json:"model_id"`
	DataSource string `json:"data_source"`
	Purpose string `json:"purpose"`
	ConsentType string `json:"consent_type"`
}

func (c CreateAgreement) CommandName() string { return "CreateAgreement" }

func (c CreateAgreement) Execute(repo DataUsageAgreementRepository) (*DataUsageAgreement, *CreatedAgreement, error) {
	agg := NewDataUsageAgreement(c.ModelId, c.DataSource, c.Purpose, c.ConsentType, time.Time{}, time.Time{}, nil, "")
	agg.Status = "draft"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := CreatedAgreement{
		AggregateID: agg.ID,
		ModelId: c.ModelId,
		DataSource: c.DataSource,
		Purpose: c.Purpose,
		ConsentType: c.ConsentType,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
